// react-push-menu/dist/index.esm.js
import React from 'react';
import styled from 'styled-components';
import clsx from 'clsx';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

function __spreadArrays() {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
}
function __makeTemplateObject(cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
}

var getNodeChildren = function (node, propMap) { return (node === null || node === void 0 ? void 0 : node[propMap === null || propMap === void 0 ? void 0 : propMap.childPropName]) || []; };
var getNodeTitle = function (node, propMap) { return node === null || node === void 0 ? void 0 : node[propMap === null || propMap === void 0 ? void 0 : propMap.displayName]; };

var defaultPropMaps = {
    displayName: 'name',
    linkClasses: 'classes',
    childPropName: 'children',
    expanderClasses: 'expClasses',
    url: 'url',
    id: 'id',
};
var Context = React.createContext({
    nodes: {},
    propMap: defaultPropMaps,
    visibleMenus: [],
    addMenu: function () { return undefined; },
    removeLastMenu: function () { return undefined; },
    closeMenu: function () { return undefined; },
    openMenu: function () { return undefined; },
    openSubMenu: function () { return undefined; },
    toggleMenu: function () { return undefined; },
});
var Provider = Context.Provider;
var PushMenuProvider = function (_a) {
    var children = _a.children, suppliedPropMap = _a.propMap, nodes = _a.nodes;
    var _b = React.useState([]), visibleMenus = _b[0], setVisibleMenus = _b[1];
    var propMap = Object.assign({}, defaultPropMaps, suppliedPropMap);
    var addMenu = function (menu) {
        setVisibleMenus(function (prevMenus) {
            return __spreadArrays(prevMenus, [menu]);
        });
    };
    var removeLastMenu = function () {
        setVisibleMenus(function (prevMenus) {
            var clonedMenus = Array.from(prevMenus);
            clonedMenus.pop();
            return clonedMenus;
        });
    };
    var closeMenu = function () {
        setVisibleMenus([]);
    };
    var openSubMenu = function (node) {
        var nodeChildren = getNodeChildren(node, propMap);
        if (nodeChildren.length < 1) {
            return null;
        }
        addMenu(node);
    };
    var openMenu = function () {
        var nodeChildren = getNodeChildren(nodes, propMap);
        if (visibleMenus.length > 0 || nodeChildren.length < 1) {
            return null;
        }
        setVisibleMenus([nodes]);
    };
    var toggleMenu = function () {
        if (visibleMenus.length > 0)
            setVisibleMenus([]);
        if (visibleMenus.length < 1)
            openMenu();
    };
    var commonProps = {
        propMap: propMap,
        nodes: nodes,
        visibleMenus: visibleMenus,
        addMenu: addMenu,
        removeLastMenu: removeLastMenu,
        closeMenu: closeMenu,
        openMenu: openMenu,
        openSubMenu: openSubMenu,
        toggleMenu: toggleMenu,
    };
    return React.createElement(Provider, { value: commonProps }, children);
};
var usePushMenu = function () { return React.useContext(Context); };

var BackItem = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  padding: 0 10px;\n\n  span svg {\n    font-size: 10px;\n  }\n"], ["\n  padding: 0 10px;\n\n  span svg {\n    font-size: 10px;\n  }\n"])));
var DefaultBackButton = function (_a) {
    var backIcon = _a.backIcon;
    var removeLastMenu = usePushMenu().removeLastMenu;
    return (React.createElement("a", { href: "#", onClick: function () { return removeLastMenu(); } },
        React.createElement(BackItem, { className: 'rpm-back-item' }, backIcon),
        React.createElement(BackItem, { className: 'rpm-back-item' }, "back")));
};
var templateObject_1;

var Link = styled.a(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject(["\n  display: inline-block;\n  width: 100%;\n  cursor: pointer;\n"], ["\n  display: inline-block;\n  width: 100%;\n  cursor: pointer;\n"])));
var DefaultLinkComponent = function (_a) {
    var data = _a.data, onNodeClick = _a.onNodeClick;
    var _b = usePushMenu(), propMap = _b.propMap, otherContextProps = __rest(_b, ["propMap"]);
    var node = data.node;
    var nodeTitle = node[propMap.displayName];
    return (React.createElement(Link, { onClick: function (e) { return onNodeClick(e, __assign(__assign(__assign({}, data), { propMap: propMap }), otherContextProps)); }, className: "rpm-node-link rpm-inline-block " + (node[propMap.linkClasses] || '') }, nodeTitle));
};
var templateObject_1$1;

var Wrapper = styled.div(templateObject_1$2 || (templateObject_1$2 = __makeTemplateObject(["\n  display: inline-block;\n  font-size: 20px;\n  cursor: pointer;\n  padding: 0.7em 1em 0.7em 1em;\n  color: white;\n\n  i {\n    top: 8px;\n  }\n\n  :hover {\n    background: rgba(0, 0, 0, 0.2);\n    box-shadow: inset 0 -1px rgba(0, 0, 0, 0);\n  }\n"], ["\n  display: inline-block;\n  font-size: 20px;\n  cursor: pointer;\n  padding: 0.7em 1em 0.7em 1em;\n  color: white;\n\n  i {\n    top: 8px;\n  }\n\n  :hover {\n    background: rgba(0, 0, 0, 0.2);\n    box-shadow: inset 0 -1px rgba(0, 0, 0, 0);\n  }\n"])));
var Expander = function (_a) {
    var data = _a.data, ExpanderComponent = _a.expanderComponent, _b = _a.onMenuExpand, onMenuExpand = _b === void 0 ? function () { return undefined; } : _b;
    var node = data.node;
    var openSubMenu = usePushMenu().openSubMenu;
    return (React.createElement(Wrapper, { className: "rpm-node-exp", onClick: function (e) {
            var allowDefault = onMenuExpand(e, __assign({ node: node }, data));
            if (typeof allowDefault === 'boolean' && !allowDefault) {
                return false;
            }
            openSubMenu(node);
        } },
        React.createElement(ExpanderComponent, { data: data })));
};
var templateObject_1$2;

var Wrapper$1 = styled.div(templateObject_1$3 || (templateObject_1$3 = __makeTemplateObject(["\n  display: flex;\n  justify-content: space-between;\n  padding: 0.7em 0;\n  font-size: 1.4em;\n"], ["\n  display: flex;\n  justify-content: space-between;\n  padding: 0.7em 0;\n  font-size: 1.4em;\n"])));
var Node = function (_a) {
    var suppliedNode = _a.node, _b = _a.linkComponent, LinkComponent = _b === void 0 ? DefaultLinkComponent : _b, expanderComponent = _a.expanderComponent, onNodeClick = _a.onNodeClick, onMenuExpand = _a.onMenuExpand;
    var menuContext = usePushMenu();
    var self = {
        renderNode: function renderNode(node) {
            var hasChildren = node.children && node.children.length > 0;
            return (React.createElement(Wrapper$1, { className: "rpm-node-cntr" },
                React.createElement(LinkComponent, { onNodeClick: onNodeClick, data: __assign({ node: node }, menuContext) }),
                hasChildren && (React.createElement(Expander, { data: __assign({ node: node }, menuContext), expanderComponent: expanderComponent, onMenuExpand: onMenuExpand }))));
        },
    };
    return self.renderNode(suppliedNode);
};
var templateObject_1$3;

var slug = function (value) {
    return value
        .replace(/\s/g, '-')
        .replace(/[()=:.,!#$@"'/\|?*+&]/g, '')
        .replace(/^-+|-+$/g, '')
        .replace(/-+/g, '-')
        .toLowerCase();
};

var Scroller = styled.div(templateObject_1$4 || (templateObject_1$4 = __makeTemplateObject(["\n  height: 100%;\n  min-height: 640px;\n  min-height: 100vh;\n  overflow-y: scroll;\n  position: relative;\n  transition: all 0.5s;\n  ", "\n"], ["\n  height: 100%;\n  min-height: 640px;\n  min-height: 100vh;\n  overflow-y: scroll;\n  position: relative;\n  transition: all 0.5s;\n  ",
    "\n"])), function (_a) {
    var open = _a.open;
    return open &&
        "\n    transform: translate3d(308px, 0px, 0px)\n  ";
});
var ScrollerInner = styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: relative;\n"], ["\n  position: relative;\n"])));
var Menu = styled.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  position: absolute; /* we can't use fixed here :( */\n  top: 0;\n  left: 0;\n  z-index: 1;\n  width: 300px;\n  height: 100%;\n\n  a {\n    text-decoration: none;\n    color: #fff;\n    width: 100%;\n    display: flex;\n    align-items: center;\n  }\n"], ["\n  position: absolute; /* we can't use fixed here :( */\n  top: 0;\n  left: 0;\n  z-index: 1;\n  width: 300px;\n  height: 100%;\n\n  a {\n    text-decoration: none;\n    color: #fff;\n    width: 100%;\n    display: flex;\n    align-items: center;\n  }\n"])));
var PusherBase = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  transition: all 0.5s;\n\n  &::after {\n    position: absolute;\n    top: 0;\n    right: 0;\n    width: 0;\n    height: 0;\n    content: '';\n    opacity: 0;\n  }\n\n  &::after {\n    background: rgba(0, 0, 0, 0.3);\n    transition: opacity 0.3s, width 0.1s 0.3s, height 0.1s 0.3s;\n  }\n"], ["\n  transition: all 0.5s;\n\n  &::after {\n    position: absolute;\n    top: 0;\n    right: 0;\n    width: 0;\n    height: 0;\n    content: '';\n    opacity: 0;\n  }\n\n  &::after {\n    background: rgba(0, 0, 0, 0.3);\n    transition: opacity 0.3s, width 0.1s 0.3s, height 0.1s 0.3s;\n  }\n"])));
var Pusher = styled(PusherBase)(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  position: relative;\n  left: 0;\n  height: 100%;\n  width: 100%;\n  opacity: 1;\n  transition: opacity 0.3s;\n"], ["\n  position: relative;\n  left: 0;\n  height: 100%;\n  width: 100%;\n  opacity: 1;\n  transition: opacity 0.3s;\n"])));
var Level = styled(PusherBase)(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  ::after {\n    z-index: -1;\n  }\n\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n  background: #336ca6;\n  padding-left: 8px;\n"], ["\n  ::after {\n    z-index: -1;\n  }\n\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n  background: #336ca6;\n  padding-left: 8px;\n"])));
var PushMenu = function (_a) {
    var children = _a.children, linkComponent = _a.linkComponent, expanderComponent = _a.expanderComponent, _b = _a.backComponent, BackComponent = _b === void 0 ? DefaultBackButton : _b, backIcon = _a.backIcon, _c = _a.onNodeClick, onNodeClick = _c === void 0 ? function () { return undefined; } : _c, _d = _a.onMenuExpand, onMenuExpand = _d === void 0 ? function () { return undefined; } : _d, _e = _a.openOnMount, openOnMount = _e === void 0 ? false : _e;
    var menuContext = usePushMenu();
    var propMap = menuContext.propMap, nodes = menuContext.nodes, visibleMenus = menuContext.visibleMenus, openMenu = menuContext.openMenu;
    var isOpen = visibleMenus.length > 0;
    React.useEffect(function () {
        if (openOnMount) {
            openMenu();
        }
    }, [openOnMount]);
    if (!isOpen) {
        return React.createElement(React.Fragment, null, children);
    }
    return (React.createElement(Pusher, { className: clsx("rpm-mp-pusher", { 'rpm-mp-pushed': isOpen }), id: "rpm-mp-pusher" },
        React.createElement(Menu, { id: "rpm-mp-menu", className: "rpm-mp-menu", open: isOpen }, visibleMenus.map(function (menuNode, level) {
            if (menuNode === void 0) { menuNode = {}; }
            var items = getNodeChildren(menuNode, propMap);
            return (React.createElement(Level, { key: level, className: "rpm-mp-level rpm-mp-level-" + level },
                React.createElement("div", null,
                    React.createElement("h2", { className: "rpm-mp-header" }, level === 0 ? nodes.header : getNodeTitle(menuNode, propMap)),
                    level > 0 && (React.createElement("div", { className: "rpm-inline-block rpm-mp-back" },
                        React.createElement(BackComponent, { data: __assign({ node: menuNode }, menuContext), backIcon: backIcon })))),
                items.map(function (node, index) {
                    var nodeTitle = getNodeTitle(node, propMap);
                    var key = (node === null || node === void 0 ? void 0 : node[propMap === null || propMap === void 0 ? void 0 : propMap.id]) || slug(nodeTitle) + "-" + index;
                    return (React.createElement(Node, { key: key, node: __assign(__assign({}, node), { nodeTitle: nodeTitle }), linkComponent: linkComponent, expanderComponent: expanderComponent, onNodeClick: onNodeClick, onMenuExpand: onMenuExpand }));
                })));
        })),
        React.createElement(Scroller, { className: "rpm-scroller", open: isOpen },
            React.createElement(ScrollerInner, { className: "rpm-scroller-inner" }, children))));
};
var templateObject_1$4, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;

var Wrapper$2 = styled.div(templateObject_1$5 || (templateObject_1$5 = __makeTemplateObject(["\n  height: 100%;\n  min-height: 640px;\n  min-height: 100vh;\n  position: relative;\n  overflow: hidden;\n"], ["\n  height: 100%;\n  min-height: 640px;\n  min-height: 100vh;\n  position: relative;\n  overflow: hidden;\n"])));
var PushMenuContainer = function (_a) {
    var nodes = _a.nodes, propMap = _a.propMap, children = _a.children, linkComponent = _a.linkComponent, expanderComponent = _a.expanderComponent, backComponent = _a.backComponent, backIcon = _a.backIcon, className = _a.className, _b = _a.onNodeClick, onNodeClick = _b === void 0 ? function () { return undefined; } : _b, _c = _a.onMenuExpand, onMenuExpand = _c === void 0 ? function () { return undefined; } : _c, _d = _a.openOnMount, openOnMount = _d === void 0 ? false : _d;
    return (React.createElement(PushMenuProvider, { propMap: propMap, nodes: nodes },
        React.createElement(Wrapper$2, { className: clsx('rpm-container', className), id: "rpm-container" },
            React.createElement(PushMenu, { linkComponent: linkComponent, expanderComponent: expanderComponent, backComponent: backComponent, backIcon: backIcon, onNodeClick: onNodeClick, onMenuExpand: onMenuExpand, openOnMount: openOnMount }, children))));
};
var templateObject_1$5;

export { DefaultBackButton, DefaultLinkComponent, PushMenu as Menu, PushMenuContainer as PushMenu, usePushMenu };
